### PR TITLE
Add studio auth to datachain

### PR DESCRIPTION
**Pull request by amritghimire**
Adds a new auth command to datachain that authorizes the datachain with
studio.

Commands added are:
- `datachain studio login`
- `datachain studio logout`
- `datachain studio token`

## Studio command
```sh
Authenticate Datachain with Studio and set the token. Once this token has been properly configured, Datachain will utilize it for seamlessly sharing
datasets and using Studio features from CLI

positional arguments:
  {login,logout,token}  Use `Datachain studio CMD --help` to display command-specific help.
    login               Authenticate Datachain with Studio host
    logout              Logout user from Studio
    token               View the token datachain uses to contact Studio

options:
  -h, --help            show this help message and exit
  --aws-endpoint-url AWS_ENDPOINT_URL
                        AWS endpoint URL
  --anon                AWS anon (aka awscli's --no-sign-request)
  --ttl TTL             Time-to-live of data source cache. Negative equals forever.
  -u, --update          Update cache
  -v, --verbose         Verbose
  -q, --quiet           Be quiet
  --debug-sql           Show All SQL Queries (very verbose output, for debugging only)
  --pdb                 Drop into the pdb debugger on fatal exception
```

## Login

To login, you can call `datachain studio login` which will open the Studio with a device code prefilled in your browser. Once the user authorizes the code, the token is saved to global config.
```sh
usage: datachain studio login [-h] [-q | -v] [-H HOSTNAME] [-s SCOPES] [-n NAME] [-d]

By default, this command authorize datachain with Studio with
 default scopes and a random  name as token name.

options:
  -h, --help            show this help message and exit
  -q, --quiet           Be quiet.
  -v, --verbose         Be verbose.
  -H HOSTNAME, --hostname HOSTNAME
                        The hostname of the Studio instance to authenticate with.
  -s SCOPES, --scopes SCOPES
                        The scopes for the authentication token.
  -n NAME, --name NAME  The name of the authentication token. It will be used to identify token shown in Studio profile.
  -d, --use-device-code
                        Use authentication flow based on user code. You will be presented with user code to enter in browser. Datachain will also use this if it
                        cannot launch browser on your behalf.
```

#### Normal flow
On running the command `datachain studio login` , following screen is presented in CLI:
```sh
datachain studio login
A web browser has been opened at
https://studio.iterative.ai/auth/device-login.
Please continue the login in the web browser.
If no web browser is available or if the web browser fails to open,
use device code flow with `datachain studio login --use-device-code`.
```

A webbrowser is opened with the device code in Studio.
image
Once user authorizes the token in the screen above, the token is saved to the user's global config.
Additional message as below will be shown in the CLI and will exit.
```sh
Authentication successful. The token will be available as risen-geum in Studio profile.
```

#### Device login flow
In case, you are running this in remote machine or where you are unable to open web browser, the following screen is presented.
```
Please open the following url in your browser.
https://studio.iterative.ai/auth/device-login
And enter the user code below 06D21JBK to authorize.
```
Every other flow is same.

## Logout
```sh
usage: datachain studio logout [-h] [-q | -v]

This command helps to log out user from  Studio.

options:
  -h, --help     show this help message and exit
  -q, --quiet    Be quiet.
  -v, --verbose  Be verbose.
```

## Token
```sh
usage: datachain studio token [-h] [-q | -v]

View the token datachain uses to contact Studio

options:
  -h, --help     show this help message and exit
  -q, --quiet    Be quiet.
  -v, --verbose  Be verbose.
```

Related to https://github.com/iterative/studio/issues/10774

Based on #513

### CODE REVIEW

**Comment by 0x2b3bfa0** [comment]
281406025-8662d52b-9523-4db4-9742-02808d7e3ec4

Why does it day "DvcX operations" instead of DataChain? 😅

File: src/datachain/cli.py [hunk 1]
```diff
@@ -97,6 +98,94 @@ def add_show_args(parser: ArgumentParser) -> None:
...
+    studio_subparser = studio_parser.add_subparsers(
+        dest="cmd",
+        help="Use `Datachain studio CMD --help` to display command-specific help.",
+        required=True,
+    )
+
+    studio_login_help = "Authenticate Datachain with Studio host"
+    studio_login_description = (
+        "By default, this command authenticates the Datachain with Studio\n"
+        " using default scopes and assigns a random name as the token name."
```
Comment (shcheklein) [inline src/datachain/cli.py:125]: minor: probably extra space at the beginning is not needed

File: src/datachain/cli.py [hunk 2]
```diff
@@ -97,6 +98,94 @@ def add_show_args(parser: ArgumentParser) -> None:
...
+        action="store_true",
+        default=False,
+        help="Use authentication flow based on user code.\n"
+        "You will be presented with user code to enter in browser.\n"
+        "Datachain will also use this if it cannot launch browser on your behalf.",
+    )
+
+    studio_logout_help = "Logout user from Studio"
+    studio_logout_description = (
+        "This removes the studio token from your global config.\n"
```
Comment (shcheklein) [inline src/datachain/cli.py:169]: do we need an extra `\n` here?

File: src/datachain/cli.py [hunk 3]
```diff
@@ -97,6 +98,94 @@ def add_show_args(parser: ArgumentParser) -> None:
     )
 
 
+def add_studio_parser(subparsers, parent_parser) -> None:
+    studio_help = "Commands to authenticate Datachain with Iterative Studio"
+    studio_description = (
+        "Authenticate Datachain with Studio and set the token."
+        " Once this token has been properly configured,\n"
+        " Datachain will utilize it for seamlessly sharing datasets\n"
```
Comment (shcheklein) [inline src/datachain/cli.py:106]: same here, review spaces and newlines

File: src/datachain/studio.py [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+
+
+def login(args: "Namespace"):
+    from dvc_studio_client.auth import StudioAuthError, get_access_token
+
+    config = Config().read().get("studio", {})
+    name = args.name
+    hostname = (
+        args.hostname
+        or os.environ.get("DATACHAIN_STUDIO_HOSTNAME")
```
Comment (shcheklein) [inline src/datachain/studio.py:34]: should we use the same DVC_STUDIO_HOSTNAME var? it's still DVC Studio

**Review by shcheklein — APPROVED** [review]
thanks @amritghimire . some minor questions here and there

**Comment by amritghimire** [comment]
> 281406025-8662d52b-9523-4db4-9742-02808d7e3ec4
> 
> 
> 
> Why does it day "DvcX operations" instead of DataChain? 😅 

Oh, I copied the description from dvc PR for the screenshot 😀

**Pull request merged**